### PR TITLE
Unfreezing request operation and unfreezing

### LINKED ISSUE #360: Implementation unfreezing request and unfreezing operation
## Background
When a client generates a unfreezing transaction, it really affects the balance of account after X blocks.
This is the unfreezing feature.

## Solution
To implement unfreezing features, two operation type is needed.
One is unfreezing request operation, and the other is unfreezing operation.

1. Client who want to unfreeze the frozen account send unfreezing request transaction.
    This transaction doesn't change the balance of the frozen account.
2. After X blocks, client who already sent unfreezing request transaction send again unfreezing transaction.
    This transaction affect the balance of the frozen account. a whole amount of balance is withdrawn from     
    frozen account.

### PR DESCRIPTION
**Pull request by soonkuk**
### Github Issue

resolves #360 

## Background
When a client generates a unfreezing transaction, it really affects the balance of account after X blocks.
This is the unfreezing feature.

## Solution
To implement unfreezing features, unfreezing request operation type is needed.

1. Client who want to unfreeze the frozen account send unfreezing request transaction.
    This transaction doesn't change the balance of the frozen account.
2. After X blocks, client who already sent unfreezing request transaction send payment transaction.
    This transaction affect the balance of the frozen account. a whole amount of balance is withdrawn from frozen account.

Make the unfreezing-request operation. and make a test for it.(isaac_simulation_unfreezing_test.go)

### CODE REVIEW

**Comment by kfangw** [comment]
@soonkuk Let us know when this PR is ready to be reviewed,
Until then, please organize the commit history.
`rebase -i HEAD~x` and `f`, `r` options are useful to do that.

**Comment by soonkuk** [comment]
@kfangw I already organized commit history about unfreezing pr. Is there more commit history to organize?

**Comment by kfangw** [comment]
@soonkuk There are a bunch of commits interleaved with other's commits.
I have no idea how to follow the commit history.

**Comment by soonkuk** [comment]
I finished this PR @kfangw @Geod24 Please review to this PR.

File: lib/node/runner/checker_ballot_transaction.go [hunk 1]
```diff
@@ -227,6 +233,10 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
 		if casted.Linked != "" && (casted.Amount%common.Unit) != 0 {
 			return errors.ErrorFrozenAccountCreationWholeUnit // FIXME
 		}
+		if casted.Linked != "" && casted.Linked != tx.Source() {
+			err = errors.ErrorFrozenAccountMustCreatedFromLinkedAccount
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Why ?

File: lib/node/runner/checker_ballot_transaction.go [hunk 2]
```diff
@@ -209,7 +213,9 @@ func ValidateTx(st *storage.LevelDBBackend, tx transaction.Transaction) (err err
 //   source = Account from where the transaction (and ops) come from
 //   tx = Transaction to check
 //
-func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op transaction.Operation) (err error) {
+func ValidateOp(st *storage.LevelDBBackend, tx transaction.Transaction, op transaction.Operation) (err error) {
+	var source *block.BlockAccount
+	source, _ = block.GetBlockAccount(st, tx.B.Source)
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Missing error handling

File: lib/node/runner/checker_ballot_transaction.go [hunk 3]
```diff
@@ -244,14 +254,79 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
 			err = errors.ErrorFrozenAccountNoDeposit
 			return
 		}
-		// If it's a frozen account, everything must be withdrawn
 		if source.Linked != "" {
+			// Source account's linked address must be equal to target address
+			if source.Linked != casted.TargetAddress() {
+				err = errors.ErrorUnfreezingToInvalidLinkedAccount
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Why ?

File: lib/node/runner/checker_ballot_transaction.go [hunk 4]
```diff
@@ -244,14 +254,79 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
...
+			}
+			// If it's a frozen account, everything must be withdrawn
 			var expected common.Amount
 			expected, err = source.Balance.Sub(common.BaseFee)
 			if casted.Amount != expected {
 				err = errors.ErrorFrozenAccountMustWithdrawEverything
 				return
 			}
+			// Unfreezing must be done after X period from unfreezing request
+			var saved []block.BlockOperation
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: For performance, we should do the check in the iteration, not allocate a new array.
Also, since it's a frozen account, there can only be 1 operation associated with this account.

File: lib/node/runner/checker_ballot_transaction.go [hunk 5]
```diff
@@ -244,14 +254,79 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
...
+			if err != nil {
+				return
+			}
+			if bo.Type != transaction.OperationUnfreezingRequest {
+				err = errors.ErrorUnfreezingRequestNotRequested
+				return
+			}
+
+			// unfreezing period is originally 241920(=12*60*24*14), but here is set to 4 for the test.
+			if lastblock.Height-bo.BlockHeight <= uint64(4) {
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: That must be a constant or variable for testing purpose, but not here in the code.

File: lib/node/runner/checker_ballot_transaction.go [hunk 6]
```diff
@@ -244,14 +254,79 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
...
+		if source.Linked != casted.TargetAddress() {
+			err = errors.ErrorUnfreezingToInvalidLinkedAccount
+		}
+		var ba *block.BlockAccount
+		ba, err = block.GetBlockAccount(st, tx.B.Source)
+		if err != nil {
+			return
+		}
+		// When unfreezing, everything must be withdrawn
+		if ba.Balance != casted.GetRequestAmount().MustAdd(common.BaseFee) {
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Don't use `MustAdd` here, it'll `panic`

File: lib/node/runner/test.go [hunk 7]
```diff
@@ -123,7 +183,7 @@ func createNodeRunnerForTesting(n int, conf *consensus.ISAACConfiguration, recv
 	}
 
 	address := kp.Address()
-	balance := common.BaseFee.MustAdd(common.BaseReserve)
+	balance := common.BaseFee.MustAdd(1000000000000)
```
Comment (Geod24) [inline lib/node/runner/test.go]: Why?

File: lib/transaction/operation.go [hunk 8]
```diff
@@ -16,6 +16,8 @@ const (
 	OperationPayment                            = "payment"
 	OperationCongressVoting                     = "congress-voting"
 	OperationCongressVotingResult               = "congress-voting-result"
+	OperationMembership                         = "membership"
```
Comment (Geod24) [inline lib/transaction/operation.go]: Doesn't look related

File: cmd/sebak/cmd/wallet/payment.go [hunk 9]
```diff
@@ -150,6 +153,7 @@ func init() {
 	PaymentCmd.Flags().BoolVar(&flagCreateAccount, "create", flagCreateAccount, "Whether or not the account should be created")
 	PaymentCmd.Flags().BoolVar(&flagFreeze, "freeze", flagFreeze, "When present, the payment is a frozen account creation. Imply --create.")
 	PaymentCmd.Flags().BoolVar(&flagDry, "dry-run", flagDry, "Print the transaction instead of sending it")
+	PaymentCmd.Flags().BoolVar(&flagUnfreezingRequest, "unfreeze-request", flagUnfreezingRequest, "Request unfreezing for frozen account")
```
[the remark this answers is not in the document]
Comment (soonkuk) [reply cmd/sebak/cmd/wallet/payment.go]: @Geod24 I think unfreeze-request is more descriptive to user.
Because user is doing request for unfreeze not unfreeze when user generate transaction.

File: cmd/sebak/cmd/wallet/payment.go [hunk 10]
```diff
@@ -242,6 +246,52 @@ func makeTransactionPayment(kpSource keypair.KP, kpDest keypair.KP, amount commo
 	return tx
 }
 
+//
+/// Make a full transaction, with a single unfreezing request operation in it
+///
+/// TODO:
+///   Move to lib (it was 'borrowed' from test code)
```
[the remark this answers is not in the document]
Comment (soonkuk) [reply cmd/sebak/cmd/wallet/payment.go]: @Geod24 I included it in mistake.
Thanks.

File: cmd/sebak/cmd/wallet/payment.go [hunk 11]
```diff
@@ -242,6 +246,52 @@ func makeTransactionPayment(kpSource keypair.KP, kpDest keypair.KP, amount commo
...
+///   kpDest   = Receiver's keypair.FromAddress address
+///   amount   = Amount to send as initial value
+///   seqid    = SequenceID of the last transaction
+///
+/// Returns:
+///  `sebak.Transaction` = The generated `Transaction` to do a unfreezing request
+///
+func makeTransactionUnfreezingRequest(kpSource keypair.KP, kpDest keypair.KP, amount common.Amount, seqid uint64) transaction.Transaction {
+	var a common.Amount
+	a = 0
```
[the remark this answers is not in the document]
Comment (soonkuk) [reply cmd/sebak/cmd/wallet/payment.go]: @Geod24 I'll make variable name more descriptively.

File: lib/node/runner/checker_ballot_transaction.go (hunk 1 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 I think this is needed for prevent mistake of user making link to other account.

File: lib/node/runner/checker_ballot_transaction.go (hunk 3 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 I think freezing and unfreezing should be executed only through linked general account.
Because I think frozen account are like as installment savings.
So isn't it natural to use freezing and unfreezing through only one account?

File: lib/node/runner/checker_ballot_transaction.go (hunk 5 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 Ok I'll make a unfreezing period constant in constant.go.

File: lib/node/runner/test.go (hunk 7 above)
Comment (soonkuk) [reply lib/node/runner/test.go]: @Geod24 For the test, there is need to create account which have more units than BaseReserve.

File: lib/transaction/operation.go (hunk 8 above)
Comment (soonkuk) [reply lib/transaction/operation.go]: @Geod24 Yes It does not related to this PR. 
I'll remove it.

File: lib/node/runner/checker_ballot_transaction.go (hunk 2 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 `ValidateOp` is called from `ValidateTx`.
In `ValidateTx` tx.B.Source is already checked before calling `ValidateOp`.
So I remove err check. Is it necessary to check tx.B.Source again?

File: cmd/sebak/cmd/wallet/payment.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Geod24) [reply cmd/sebak/cmd/wallet/payment.go]: Okay, it's not such an important point, so let's leave it as is.

File: lib/node/runner/checker_ballot_transaction.go (hunk 1 above)
Comment (Geod24) [reply lib/node/runner/checker_ballot_transaction.go]: The node itself should not enforce this. because it has nothing to do with the ability of the node to make consensus. There is nothing definitely wrong with a frozen account benefitting someone else. If we want to enforce such a thing, it should be done client side.

Moreover, I disagree that the mistake is easy to make. The user has to put a valid account address in there. So simple errors like typos will get caught.

File: lib/node/runner/checker_ballot_transaction.go (hunk 3 above)
Comment (Geod24) [reply lib/node/runner/checker_ballot_transaction.go]: I don't think we should put that restriction in place. We should only care about things that are definitely wrong, not trying to install policies. If account A1 decides to make frozen account F1, which rewards A2, and then to give the key to A3, that's his decision.

File: lib/node/runner/checker_ballot_transaction.go (hunk 1 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 Ok I agree that it should be done client side. I remove it.

File: lib/node/runner/checker_ballot_transaction.go [hunk 12]
```diff
@@ -282,14 +288,71 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
...
+			if err != nil {
+				return
+			}
+			if bo.Type != transaction.OperationUnfreezingRequest {
+				err = errors.ErrorUnfreezingRequestNotRequested
+				return
+			}
+
+			// unfreezing period is 241920.
+			if lastblock.Height-bo.BlockHeight <= common.UnfreezingPeriod {
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Shouldn't it be `<` ?

File: lib/node/runner/checker_ballot_transaction.go (hunk 2 above)
Comment (Geod24) [reply lib/node/runner/checker_ballot_transaction.go]: Oh wait, this function was previously taking the `BlockAccount` as input, so why moving it inside ?
Also, I don't think `tx` is necessary anymore, is it ?

File: lib/node/runner/checker_ballot_transaction.go [hunk 13]
```diff
@@ -282,14 +288,71 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
...
+			err = errors.ErrorUnfreezingFromInvalidAccount
+			return
+		}
+		// Target account must be existed
+		var exists bool
+		if exists, err = block.ExistsBlockAccount(st, casted.TargetAddress()); err == nil && !exists {
+			err = errors.ErrorBlockAccountDoesNotExists
+			return
+		}
+		// Source account's linked address must be equal to target address
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Why ? I don't think that's necessary either

File: lib/node/runner/checker_ballot_transaction.go [hunk 14]
```diff
@@ -282,14 +288,71 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
...
+				return
+			}
+
+			// unfreezing period is 241920.
+			if lastblock.Height-bo.BlockHeight <= common.UnfreezingPeriod {
+				err = errors.ErrorUnfreezingNotReachedExpiration
+				return
+			}
+		}
+	case transaction.OperationUnfreezingRequest:
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: The original design (which we discussed) was that the "unfreezing request" would just be a payment operation.
That will allow you to reuse most of this function with much less code.

File: lib/node/runner/test.go (hunk 7 above)
Comment (Geod24) [reply lib/node/runner/test.go]: The only difference between this function and `createNodeRunnerForTesting` is the initial balance, isn't it ?
Then just make the initial balance a parameter. Another possibility is to reduce the initial balance of account, as I assume your problem is that unfreezing an account will go over the `MaximumBalance`.

Also, I asked "why" because the amount looks rather arbitrary. It does not related to `BaseReserve` in any way.
If you used something like `common.BaseReserve.MustAdd(common.BaseFee)` it would be more explicit for example.

File: lib/node/runner/checker_ballot_transaction.go (hunk 2 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 Yes you are right. I'll fix it.

File: lib/node/runner/checker_ballot_transaction.go (hunk 12 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 Yes you are right. I was mistaken. Thanks.

File: lib/node/runner/checker_ballot_transaction.go (hunk 14 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 Do you mean if some user wants unfreezing, he only need to send two payment txs at X blockheights interval?

File: lib/node/runner/checker_ballot_transaction.go (hunk 13 above)
Comment (soonkuk) [reply lib/node/runner/checker_ballot_transaction.go]: @Geod24 I'll remove it. Thanks.

**Comment by kfangw** [comment]
defer to @Geod24

**Comment by kfangw** [comment]
defer to @Geod24

**Comment by Geod24** [comment]
@soonkuk : Can you rebase on master ?

**Comment by soonkuk** [comment]
@Geod24 Ok I rebased it.

File: lib/transaction/operation/unfreezing_request.go [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+	"boscoin.io/sebak/lib/common"
+)
+
+type UnfreezeRequest struct {
+	Target        string        `json:"target"`
+	Amount        common.Amount `json:"amount"`
+	RequestAmount common.Amount `json:"request-amount"`
+}
+
+func NewUnfreezeRequest(target string, amount common.Amount, request_amount common.Amount) UnfreezeRequest {
```
Comment (zzim2x) [inline lib/transaction/operation/unfreezing_request.go]: RequestAmount value is not necessary in an unfreeze request. To be unfreeze amount is all or nothing.

Comment (Geod24) [reply lib/transaction/operation/unfreezing_request.go]: We discussed it with @soonkuk . While it's not necessary, if we make it a payment operation, it's not possible to differentiate when the request is received whether or not it's a normal payment or an unfreezing, which means we'd have to move one check from `IsWellFormed` to `ValidateOp`. I don't think it's worth it, since unfreezing requests are so rare.

Comment (zzim2x) [reply lib/transaction/operation/unfreezing_request.go]: @soonkuk That's right. 

Anyway, I don't think `UnfreezeRequest` struct should be a Payable operation. The amount value is always 0 and target is not required, it is just a request.

All frozen accounts are able to send a tx including a payable operation and make unfreeze requests repeatedly. These operations are not blocked yet.

**Comment by soonkuk** [comment]
@zzim2x I revised the code. unfreeze request operation only need target address and node blocks repeated request.

File: cmd/sebak/cmd/wallet/unfreezeRequest.go [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+///
+/// Params:
+///   conn = Network connection to the node to request
+///   sender = sender from which to request the account (only `Address` is used)
+///
+/// Returns:
+///   sebak.BlockAccount = The deserialized block account, or a default-initialized one if an error occured
+///   error = `nil` or the error that occured (either network or deserialization)
+///
+func getFrozenAccountDetails(conn *network.HTTP2NetworkClient, frozenAccount keypair.KP) (block.BlockAccount, error) {
```
Comment (Geod24) [inline cmd/sebak/cmd/wallet/unfreezeRequest.go]: How is this different from `getAccountDetails` ?

File: lib/common/constant.go [hunk 17]
```diff
@@ -26,6 +26,11 @@ const (
 
 	// BlockHeightEndOfInflation sets the block height of inflation end.
 	BlockHeightEndOfInflation uint64 = 36000000
+	// UnfreezingPeriod is the blockheight duration for unfreezing.
+	// Frozen account can be unfreezed after passing unfreezing period from unfreezing request.
+	// It can be calculated like this. 241920 = 12*60*24*14. This period is considered as about two weeks.
+	// For test current value is 4
+	UnfreezingPeriod uint64 = 4
```
Comment (Geod24) [inline lib/common/constant.go]: That should probably be changed, no ?

File: lib/node/runner/checker.go [hunk 18]
```diff
@@ -731,3 +737,15 @@ func finishInflation(st *storage.LevelDBBackend, opb operation.Inflation, log lo
 
 	return
 }
+
+func finishUnfreezeRequest(st *storage.LevelDBBackend, tx transaction.Transaction, opb operation.UnfreezeRequest, log logging.Logger) (err error) {
```
Comment (Geod24) [inline lib/node/runner/checker.go]: This does not do anything, does it ?

File: lib/common/constant.go [hunk 19]
```diff
@@ -26,6 +26,11 @@ const (
 
 	// BlockHeightEndOfInflation sets the block height of inflation end.
 	BlockHeightEndOfInflation uint64 = 36000000
+	// UnfreezingPeriod is the blockheight duration for unfreezing.
```
Comment (Geod24) [inline lib/common/constant.go]: UnfreezingPeriod defines the number of blocks required for unfreezing

File: lib/transaction/test.go [hunk 20]
```diff
@@ -87,3 +87,93 @@ func MakeTransactionCreateAccount(kpSource *keypair.Full, target string, amount
...
+		H: operation.Header{
+			Type: operation.TypeCreateAccount,
+		},
+		B: opb,
+	}
+
+	txBody := Body{
+		Source:     kpSource.Address(),
+		Fee:        common.BaseFee,
+		SequenceID: rand.Uint64(),
```
Comment (Geod24) [inline lib/transaction/test.go]: That won't be very useful

File: lib/node/runner/checker.go (hunk 18 above)
Comment (soonkuk) [reply lib/node/runner/checker.go]: @Geod24 Yes, it check only existence of source account.

File: lib/node/runner/checker_ballot_transaction.go [hunk 21]
```diff
@@ -294,14 +319,52 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op opera
...
+			}
+			// unfreezing period is 241920.
+			if lastblock.Height-bo.BlockHeight < common.UnfreezingPeriod {
+				err = errors.ErrorUnfreezingNotReachedExpiration
+				return
+			}
+		}
+	case operation.TypeUnfreezingRequest:
+		var ok bool
+		if _, ok = op.B.(operation.UnfreezeRequest); !ok {
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: `ok` doesn't have to be declared outside of this scope

File: lib/node/runner/checker_ballot_transaction.go [hunk 22]
```diff
@@ -294,14 +319,52 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op opera
...
+			err = errors.ErrorTypeOperationBodyNotMatched
+			return
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: `return errors. ErrorTypeOperationBodyNotMatched`

File: lib/node/runner/checker.go (hunk 18 above)
Comment (Geod24) [reply lib/node/runner/checker.go]: But that's already done before, isn't it ?

Comment (soonkuk) [reply lib/node/runner/checker.go]: @Geod24 Yes that's right. I leaved only log.

File: lib/common/constant.go (hunk 19 above)
Comment (soonkuk) [reply lib/common/constant.go]: @Geod24 Yes I changed it.

**Comment by soonkuk** [comment]
@Geod24 I revised codes. Please check :).

**Comment by soonkuk** [comment]
@Geod24 Can you check this PR?

File: lib/common/constant.go (hunk 19 above)
Comment (Geod24) [reply lib/common/constant.go]: I don't see the change ?

File: lib/network/api/account.go [hunk 23]
```diff
@@ -16,7 +16,6 @@ import (
 func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	address := vars["id"]
-
```
Comment (Geod24) [inline lib/network/api/account.go]: Unrelated

File: lib/network/api/operation.go [hunk 24]
```diff
@@ -51,6 +53,7 @@ func (api NetworkHandlerAPI) GetOperationsByAccountHandler(w http.ResponseWriter
 
 	txs := readFunc() //TODO paging support
 	self := r.URL.String()
+
```
Comment (Geod24) [inline lib/network/api/operation.go]: Changes in this file are unrelated

File: lib/node/runner/checker.go [hunk 25]
```diff
@@ -184,8 +184,8 @@ func BallotAlreadyVoted(c common.Checker, args ...interface{}) (err error) {
 // and the round of ballot is not yet registered, this will make new
 // `RunningRound`.
 func BallotVote(c common.Checker, args ...interface{}) (err error) {
-	checker := c.(*BallotChecker)
 
+	checker := c.(*BallotChecker)
```
Comment (Geod24) [inline lib/node/runner/checker.go]: Unrelated

File: lib/node/runner/checker.go [hunk 26]
```diff
@@ -662,6 +668,7 @@ func finishOperationPayment(st *storage.LevelDBBackend, tx transaction.Transacti
 
 func finishProposerTransaction(st *storage.LevelDBBackend, blk block.Block, ptx ballot.ProposerTransaction, log logging.Logger) (err error) {
 	{
+
```
Comment (Geod24) [inline lib/node/runner/checker.go]: Unrelated

File: lib/node/runner/checker_ballot_transaction.go [hunk 27]
```diff
@@ -265,51 +269,95 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op opera
...
+			// Unfreezing must be done after X period from unfreezing request
+			iterFunc, closeFunc := block.GetBlockOperationsBySource(st, source.Address, nil)
+			bo, _, _ := iterFunc()
+			closeFunc()
+			// Before unfreezing payment, unfreezing request shoud be saved
+			if bo.Type != operation.TypeUnfreezingRequest {
+				return errors.ErrorUnfreezingRequestNotRequested
+			}
+			var lastblock block.Block
+			lastblock, err = block.GetLatestBlock(st)
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: This can be updated now

File: lib/node/runner/checker_ballot_transaction.go [hunk 28]
```diff
@@ -265,51 +269,95 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op opera
...
```
Comment (Geod24) [inline lib/node/runner/checker_ballot_transaction.go]: Ditto

**Comment by soonkuk** [comment]
@Geod24 I removed unrelated modification. Please check. I appreciate for your detailed reviews.

**Comment by soonkuk** [comment]
@Geod24 Can you check this PR?

File: lib/node/runner/test.go [hunk 29]
```diff
@@ -49,6 +49,65 @@ func GetTransaction() (transaction.Transaction, []byte) {
...
+	tx.Sign(kpSource, networkID)
+
+	if txByte, err := tx.Serialize(); err != nil {
+		panic(err)
+	} else {
+		return tx, txByte
+	}
+}
+
+func TestGenerateNewSequenceID() uint64 {
```
Comment (Geod24) [inline lib/node/runner/test.go]: Superfluous
